WoWLan settings

alfonsosanchezbeato opened the pull request:
This is a follow-up of https://mail.gnome.org/archives/networkmanager-list/2017-January/msg00020.html

This MP add supports for wake-on-wlan, modeled after the existing support for wake-on-lan.

Besides the connection settings, wake-on-wlan itself is configured in the end via nl80211 configuration messages.

vhumpa commented:
NM CI: Please confirm test execution over this PR. Comment .*ok\W+to\W+test.* to confirm and/or .*add\W+to\W+whitelist.* to add the PR author in whitelist for future.

alfonsosanchezbeato commented:
I am not sure why the meson build has failed, I have tried locally and things work as expected. autogen works well in CI though.

alfonsosanchezbeato commented on src/platform/wifi/wifi-utils-nl80211.c:
```diff
@@ -273,6 +273,45 @@ wifi_nl80211_set_powersave (WifiData *data, guint32 powersave)
+
+	nla_nest_end(msg, triggers);
+
+	err = nl80211_send_and_recv (nl80211, msg, NULL, NULL);
```
Good catch. I have actually seen a couple of additional places where `nlmsg_free` was unnecessarily called and removed that.

alfonsosanchezbeato commented:
@thom311 thanks for the comments, PR refreshed after addressing them.

alfonsosanchezbeato commented:
Changed to fix the meson build. "semaphoreci" still fails though, with error:

The generated file "./clients/common/settings-docs.h" differs from the source file "./clients/common/settings-docs.h.in". You probably should copy the generated file over to the source file. You can skip this test by setting $NMTST_NO_CHECK_SETTINGS_DOCS=yes. You can also automatically copy the file by rerunning the test with $NM_TEST_REGENERATE=1
make[3]: *** [check-local-settings-docs] Error 1

This looks more like an issue with CI that with the changes, as I get the same error when running locally the tests, but if I run:

NM_TEST_CLIENT_CHECK_L10N=1 NM_TEST_REGENERATE=1 make -j6 check

the command is successful.

lkundrak commented:
A `NEWS` file entry for this would be very nice.

alfonsosanchezbeato commented:
@lkundrak done (and rebased)

thom311 commented on src/devices/wifi/nm-device-wifi.c:
```diff
@@ -2421,6 +2421,45 @@ build_supplicant_config (NMDeviceWifi *self,
+		                                    NM_SETTING_WIRELESS_WAKE_ON_WLAN_DEFAULT);
+
+		if (   NM_FLAGS_ANY (wowl, NM_SETTING_WIRELESS_WAKE_ON_WLAN_EXCLUSIVE_FLAGS)
+		    && !nm_utils_is_power_of_two (wowl)) {
```
This is the same check as in `verify()`. Same comment about invalid flags applies, and what happens if MAGIC is present (but obviously, no password).

thom311 commented on libnm-core/nm-setting-wireless.c:
```diff
@@ -885,6 +889,34 @@ verify (NMSetting *setting, NMConnection *connection, GError **error)
 	}
 
+	if (   NM_FLAGS_ANY (priv->wowl, NM_SETTING_WIRELESS_WAKE_ON_WLAN_EXCLUSIVE_FLAGS)
+	    && !nm_utils_is_power_of_two (priv->wowl)) {
```
should also regject large flags `!NM_FLAGS_ANY (priv->wowl, NM_SETTING_WIRELESS_WAKE_ON_WLAN_ALL)`.

thom311 commented on libnm-core/nm-setting-wireless.h:
```diff
@@ -57,6 +99,8 @@ G_BEGIN_DECLS
 #define NM_SETTING_WIRELESS_POWERSAVE   "powersave"
 #define NM_SETTING_WIRELESS_MAC_ADDRESS_RANDOMIZATION   "mac-address-randomization"
+#define NM_SETTING_WIRELESS_WAKE_ON_WLAN "wake-on-wlan"
+#define NM_SETTING_WIRELESS_WAKE_ON_WLAN_PASSWORD "wake-on-wlan-password"
```
this name is a bit unfortunate, because it has the same prefix as `NMSettingWirelessWakeOnWLan` flags. Maybe the enum should be called `NMWirelessWakeOnWLanFlags`, and the values `NM_WIRELESS_WAKE_ON_WLAN_FLAGS_*`. What do others thing?

thom311 commented on src/nm-manager.c:
```diff
@@ -1567,14 +1573,19 @@ remove_device (NMManager *self,
-	        allow_unmanage, nm_device_get_managed (device, FALSE));
+	_LOG2D (LOGD_DEVICE, device, "removing device (allow_unmanage %d, managed %d, wol %d)",
+	        allow_unmanage, nm_device_get_managed (device, FALSE),
+			device_is_wake_on_lan (priv->platform, device));
```
indention

thom311 commented on clients/common/nm-meta-setting-desc.c:
```diff
@@ -7252,6 +7252,31 @@ static const NMMetaPropertyInfo *const property_infos_WIRELESS[] = {
+				.value_infos =          ENUM_VALUE_INFOS (
+					{
+						.value = NM_SETTING_WIRELESS_WAKE_ON_WLAN_NONE,
+						.nick = "none",
```
note, that `value_infos` is only used during set, that is, when parsing the string value. Unlike the `value_infos_get` field which is used by the getter. With set-only specifying `"none"` is redundant to what we get from the GEnumClass -- it would make sense with the getter, to ensure `"none"` has highest priority.

thom311 commented on clients/common/nm-meta-setting-desc.c:
```diff
@@ -7252,6 +7252,31 @@ static const NMMetaPropertyInfo *const property_infos_WIRELESS[] = {
+					},
+					{
+						.value = NM_SETTING_WIRELESS_WAKE_ON_WLAN_NONE,
+						.nick = "disabled",
```
Nowadays, I am not convinced about aliases. There should be few ways to achieve something. Often aliases exist because of historic reasons. Ok, maybe "disabled" is a good alias, but do we really need "disable" and "disabled"? Maybe just choose one.

thom311 commented on clients/common/settings-docs.h.in:
```diff
@@ -21,6 +21,8 @@
 #define DESCRIBE_DOC_NM_SETTING_WIRELESS_SSID N_("SSID of the Wi-Fi network. Must be specified.")
 #define DESCRIBE_DOC_NM_SETTING_WIRELESS_TX_POWER N_("If non-zero, directs the device to use the specified transmit power. Units are dBm.  This property is highly driver dependent and not all devices support setting a static transmit power.")
+#define DESCRIBE_DOC_NM_SETTING_WIRELESS_WAKE_ON_WLAN N_("The NMSettingWirelessWakeOnWLan options to enable. Not all devices support all options. May be any combination of NM_SETTING_WIRELESS_WAKE_ON_WLAN_ANY (0x2), NM_SETTING_WIRELESS_WAKE_ON_WLAN_DISCONNECT (0x4), NM_SETTING_WIRELESS_WAKE_ON_WLAN_MAGIC (0x8), NM_SETTING_WIRELESS_WAKE_ON_WLAN_GTK_REKEY_FAILURE (0x10), NM_SETTING_WIRELESS_WAKE_ON ...
+#define DESCRIBE_DOC_NM_SETTING_WIRELESS_WAKE_ON_WLAN_PASSWORD N_("If specified, the password used with magic-packet-based Wake-on-WLAN, represented as a MAC address.  If NULL, no password will be required.")
```
Could you squash this commit into `libnm-core: add wake-on-wlan configuration items`, otherwise, `make check` will break for the intermediate commits.

Note: you can regenerate the file with `NM_TEST_REGENERATE=1 make check`. Note however, to properly generate all files in `clients/tests/test-client.check-on-disk/`, you have to type `make install` before running the tests (so that localization files are installed in your `--prefix`). Essentially:

```
./configure --prefix=$SOMEWHERE
make && make install
NM_TEST_REGENERATE=1 make -j 10 check
```
(or just don't commit the changes under `clients/tests`).

bengal commented:
Hi, I'm missing some context about this patchset... can you please
explain in the commit message what is the big picture of what you want
to achieve, and how it was tested.

In commit "don't shutdown interfaces if they have wowlan enabled", we
don't deconfigure devices (Wi-Fi and Ethernet) with wake-on-(w)lan active.
This means that the logic in unmanaged_on_quit() is overridden (also
for Ethernet devices). Also, NMDeviceWifi.unmanaged_on_quit() returns
TRUE for some reasons explained in comments, why Wake-on-WLAN should
make a difference when NM is stopped?

I guess a possible explanation (?) is that on shutdown we want to keep
the device configured so that it can wake up the system. But won't
wpa_supplicant bring down the interface anyway when it is terminated?

alfonsosanchezbeato commented:
@bengal I probably need to explain this better in the commit, but, to summarize, this is for handling the hibernate/resume case. NM is stopped in that case, but we need the wifi connection/IP to be kept alive so it can receive packages for WoWLAN to be triggered.

wpa_supplicant has the same issue and I have a patch for it that I will submit, with the same behaviour: if WoWLAN is enabled, the interface is not deconfigured.

alfonsosanchezbeato commented on libnm-core/nm-setting-wireless.c:
```diff
@@ -885,6 +889,34 @@ verify (NMSetting *setting, NMConnection *connection, GError **error)
+		return FALSE;
+	}
+
+	if (priv->wowl_password && !NM_FLAGS_HAS (priv->wowl, NM_SETTING_WIRELESS_WAKE_ON_WLAN_MAGIC)) {
```
Yes, password is optional.

alfonsosanchezbeato commented on src/devices/wifi/nm-device-wifi.c:
```diff
@@ -2421,6 +2421,45 @@ build_supplicant_config (NMDeviceWifi *self,
+		if (wowl != NM_SETTING_WIRELESS_WAKE_ON_WLAN_DEFAULT)
+			goto found;
+	}
+	wowl = NM_SETTING_WIRELESS_WAKE_ON_WLAN_IGNORE;
```
I have changed `wifi_nl80211_set_wake_on_wlan()` so it checks NM_SETTING_WIRELESS_WAKE_ON_WLAN_IGNORE. This ways things are handled symmetrically to the WoLAN case.

alfonsosanchezbeato commented:
@thom311 thanks for the review. I have repushed after addressing your changes.

One thing that I have done in the new patchset is removing the `wake_on_wlan_password` setting. This was added for symmetry with the WoLAN settings, but after a closer look that did not made much sense. The nl802.11 interface does not have direct support for a password, but for a more generic trigger called `NL80211_WOWLAN_TRIG_WAKEUP_PKT_80211`, which can be configured to happen for arbitrary package payloads. As adding full support for that could get a bit complex, I prefer to propose this change with support for the more basic triggers, and propose later another PR that will take care of the former.

alfonsosanchezbeato commented:
Re: testing, it can be checked that the triggers are set with a sequence like:

$ sudo nmcli c modify <conn_name> 802-11-wireless.wake-on-wlan magic
$ sudo nmcli c down <conn_name>
$ sudo nmcli c up <conn_name>
$ iw phy phy0 wowlan show
WoWLAN is enabled:
 * wake up on magic packet

For testing if the commit that leaves the wifi interface configured works, it is a matter of stopping the NM service and checking that the interface is still configured.

thom311 commented:
Pushed two fixes to th/review/pr/124 branch. Would you agree to them, and incorporate them? (either just take as separate commits, or rebase into the orignal patches, as you prefer).

Only question I have left is:

It seems to me, if you deactive a profile that set wake on wlan during activation, shouldn't it restore the device's state that it found before activation? That would involve retrieving the value first, and cache it.

alfonsosanchezbeato commented:
@thom311 I have taken your commits, thanks.

About restoring previous wowlan settings on profile deactivation, I am not sure on whether NM should restore the previous state or simply leave the driver defaults. What is done for WoLAN? Where should this be changed?

lkundrak commented:
>> It seems to me, if you deactive a profile that set wake on wlan during activation, shouldn't it restore the device's state that it found before activation? That would involve retrieving the value first, and cache it.

> About restoring previous wowlan settings on profile deactivation, I am not sure on whether NM should restore the previous state or simply leave the driver defaults. What is done for WoLAN? Where should this be changed?

I'm not sure what exactly should be done. We probably want to keep the WoWLAN setting when the connection is being brought down because the machine is being suspended, shut down or the connection failed (AP become unavailable or addresses expired because NDP or DHCP stopped responding).

I'd suggest that we turn off WoWLAN only when the device changes state due to NM_DEVICE_STATE_REASON_USER_REQUESTED. Note that I say turned off -- I'm not convinced we necessarily need to restore the previous state (I'd be fine with it, but it seems to be just extra hassle to me).

(Note that we don't seem to be doing this for wired WOL -- I think that's a bug.)

alfonsosanchezbeato commented:
@lkundrak branch repushed after addressing your comments. I have added a commit that removes WoWLAN configuration when the connection is teared down. That path is not followed when NM terminates or before going to S3, so it does not interfere with leaving WoWLAN active in that case. However, I think it should be better do this only for `reason==NM_DEVICE_STATE_REASON_USER_REQUESTED`, but for that I think I new parameter would be needed in device->deactivate()

alfonsosanchezbeato commented on src/nm-manager.c:
```diff
@@ -1558,6 +1558,12 @@ _parent_notify_changed (NMManager *self,
 	}
 }
 
+static gboolean
```
Added to commit message. FTR http://lists.infradead.org/pipermail/hostap/2018-June/038644.html

thom311 commented:
Merged to master as b29bf1b0791d5ef5cc111d629fe7a01e5f13fec3

The last commit was not included here. Since, WakeOnLan has the same problem... it would be great however, to fix that still.

alfonsosanchezbeato commented:
@thom311 awesome! Closing the PR then.

thom311 commented:
Let's re-open. The last patch still addresses a valid point and needs adjustment :)

vhumpa commented:
NM CI: Please confirm test execution over this PR. Comment .*ok\W+to\W+test.* to confirm and/or .*add\W+to\W+whitelist.* to add the PR author in whitelist for future.

thom311 commented:
lgtm

bengal commented on src/platform/wifi/wifi-utils-nl80211.c:
```diff
@@ -271,6 +271,69 @@ wifi_nl80211_set_powersave (WifiData *data, guint32 powersave)
+	nl80211_get_wake_on_wlan (nl80211, &wowl);
+
+	return wowl;
+}
```
I think the two functions above can be merged. Otherwise, LGTM.

thom311 commented:
merged as 41a784d7f16c7e71cdf74b8b22b2efc954464762 with a follow-up patch.

Thanks Alfonso!!
